base/usr/share/man/man1/ls.1
.TH LS 1 "July, 2026"

.SH NAME
ls \- list directory contents

.SH SYNOPSIS
\fBls\fP [\fB\-a\fP|\fB\-A\fP] [\fB\-f\fP] [\fB\-F\fP|\fB\-p\fP] [\fB\-h\fP]
[\fB\-i\fP] [\fB\-k\fP] [\fB\-s\fP] [\fB\-l\fP|\fB\-x\fP|\fB\-C\fP|\fB\-1\fP]
[\fB\-L\fP|\fB\-H\fP|\fB\-P\fP] [\fIpath\fP]...

.SH DESCRIPTION
List information about the files in a directory, sorted by name.

.SS Options

.TP
.BR \-a " or " \-\-all
Show hidden files (ones starting with .)

.TP
.BR \-A " or " \-\-almost\-all
Show hidden files, but not '.' or '..'.

.TP
.B \-f
Display entries in their original order as returned from the file system, instead of sorting them.

.TP
.BR \-F " or " \-\-classify
Print a symbol next to file names to identify directories (/), FIFOs (|), sockets (=),
executables (*), and symbolic links (@).

.TP
.BR \-h " or " \-\-human\-readable
When printing file sizes, use human-readable values like \fB2M\fP.
Also applies to the block sizes from the \fB\-s\fP option.

.TP
.B \-H
Arguments provided on the commandline which refer to symbolic links
to directories will be followed (if the \fBstat\fP call made after
the initial \fBlstat\fP call for a commandline argument indicates
that the argument points to a directory, the \fBstat\fP result
will be used instead of the \fBlstat\fP result).
This is the default unless \fB\-l\fP or \fB\-F\fP is enabled,
and can otherwise be disabled with \fB\-P\fP.

.TP
.BR \-i " or " \-\-inode
Display the serial number of each file before its name.

.TP
.BR \-k " or " \-\-kibibytes
Does nothing (block sizes are always 1024 bytes).
Included for compatibility.

.TP
.B \-l
Format each file on one line, with details.

.TP
.BR \-L " or " \-\-dereference
Follow symbolic links.
Effectively switches all \fBlstat\fP calls with \fBstat\fP calls.

.TP
.B \-p
Display a '/' after the name of a directory.

.TP
.B \-P
Do not follow symbolic links.
Cancels the effects of \fB\-H\fP and \fB\-L\fP.
Symbolic links will still be \fBstat\fP-ed to determine how to color
them or to print their targets when using \fB\-l\fP.

.TP
.BR \-s " or " \-\-size
Display the block size, in units of 1024 bytes, before each file name.
(As of writing, this number is based on \fBst_size\fP, rather than an
actual block count reported by the file system.)

.TP
.B \-x
Format output into lines, rather than columns. This is the old default
prior to ToaruOS 2.3.

.TP
.B \-C
Format output into columns. This should match the formatting of most
other OSes, though we do not perform smart column sizing like GNU coreutils.

.TP
.B \-\-color\fR[=\fP\fIWHEN\fP\fR]\fP
Enable color.
With \fIWHEN\fP, either \fBalways\fP, \fBnever\fP, or \fBauto\fP.
\fBauto\fP is the default and will enable color when the output is a terminal.

The \fBLS_COLORS\fP environment variable can be used to configure the colors, with mostly the same format as GNU.

.SH EXIT STATUS
.IP 0
Normally.
.IP 1
If a requested path could not be accessed or some other error occurred.

.SH ENVIRONMENT

.TP
.B LS_COLORS
A colon-separated list of \fItype\fP=\fIcolor\fP pairs.
.sp
Escape sequences are supported in both \fItype\fP and \fIcolor\fP.
.sp
The \fItype\fP may be a \fB*\fP followed by a string to match at the end of a file name, or one of the following special values:
.RS

.TP
.B lc
The sequence to print before each color, eg. "\fB\\033[\fP".

.TP
.B rc
The sequence to print after each color, eg. "\fBm\fP".

.TP
.B rs
The color string to use to reset the color, eg. "\fB0\fP".

.TP
.B ec
The full sequence printed after colored text.
This should generally be left unset, and the sequence \fBlc rs rc\fP will be used.

.TP
.B di
Directory.

.TP
.B ln
Symbolic link.
As a special case, if the \fIcolor\fP is set to "\fBtarget\fP", symbolic links will use the coloring of their targets.

.TP
.B bd
Block device.

.TP
.B cd
Character device.

.TP
.B or
Dangling symbolic link.

.TP
.B ex
Executable.

.TP
.B su
Set-user-ID file.

.TP
.B sg
Set-group-ID file.

.TP
.B pi
Pipe.

.TP
.B so
Socket.

.TP
.B no
Normal text (currently unused).

.TP
.B fi
Regular file. Used if no other specification matches.

.TP
.B mi
Missing target of a dangling symbolic link.

.TP
.B st
Directory with sticky bit set.

.TP
.B ow
Directory with \fBw+o\fP.

.TP
.B tw
Directory with \fBw+o\fP and sticky bit.
.RE

.TP
.B LS_COLSEP
Set the minimum separation between columns.
The default is 1.

.SH AUTHOR
This \fIls\fP was written by K. Lange <klange@toaruos.org>


apps/ls.c
/**
 * @brief List files
 *
 * Lists files in a directory, with nice color
 * output like any modern ls should have.
 *
 * @copyright
 * This file is part of ToaruOS and is released under the terms
 * of the NCSA / University of Illinois License - see LICENSE.md
 * Copyright (C) 2013-2026 K. Lange
 */
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <dirent.h>
#include <termios.h>
#include <time.h>
#include <pwd.h>
#include <wchar.h>
#include <getopt.h>
#include <errno.h>
#include <err.h>

#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/time.h>

#include <toaru/list.h>
#include <toaru/decodeutf8.h>
#include <toaru/lscolors.h>

#define MAX(a, b) (((a) > (b)) ? (a) : (b))

static int human_readable = 0;
static int stdout_is_tty = 1;
static int this_year = 0;
static int show_hidden = 0;
static int long_mode   = 0;
static int print_dir   = 0;
static int term_width = 0;
static int columns = 1;
static int in_order = 0;
static int show_inode = 0;
static int show_size = 0;
static int one_column = 0;
static int show_slash = 0;
static int use_color = 0;
static int min_col_spacing = 1;
static int use_sym_target = 0;
static int follow_cmdline_dir_links = 0;
static int follow_all_links = 0;

struct tfile {
	char * name;
	struct stat statbuf;
	char * link;
	struct stat statbufl;
	int lstatres;
};

static int filecmp(const void * c1, const void * c2) {
	const struct tfile * d1 = *(const struct tfile **)c1;
	const struct tfile * d2 = *(const struct tfile **)c2;

	int a = S_ISDIR(d1->statbuf.st_mode);
	int b = S_ISDIR(d2->statbuf.st_mode);

	if (a == b) return strcmp(d1->name, d2->name);
	else if (a < b) return -1;
	else if (a > b) return 1;
	return 0; /* impossible ? */
}

static int filecmp_notypesort(const void * c1, const void * c2) {
	const struct tfile * d1 = *(const struct tfile **)c1;
	const struct tfile * d2 = *(const struct tfile **)c2;

	return strcmp(d1->name, d2->name);
}

static int display_width_of_string(const char * str) {
	int out = 0;
	uint32_t c, state = 0;
	for (uint8_t *s = (uint8_t*)str; *s; s++) {
		if (!decode(&state, &c, *s)) out += wcwidth(c);
		else if (state == UTF8_REJECT) state = 0;
	}
	return out;
}

static size_t as_1k_blocks(off_t size) {
	size_t blocks = size / 1024;
	if (size % 1024) blocks += 1;
	return blocks;
}

static int print_human_readable_size(char * _out, size_t avail, size_t s);

static void prefixes(int *colwidth, struct tfile * file) {
	/* Prefixed stuff */
	if (show_inode) printf("%*zu ", colwidth[1]-1, file->statbuf.st_ino);
	if (show_size) {
		size_t size = as_1k_blocks(file->statbuf.st_size);
		if (human_readable) {
			char tmp[100];
			print_human_readable_size(tmp, 100, size * 1024);
			printf("%*s ", colwidth[2]-1, tmp);
		} else {
			printf("%*zu ", colwidth[2]-1, size);
		}
	}
}

static void print_classifier(struct stat *statbuf, int hide_link) {
	if (show_slash) {
		if (S_ISDIR(statbuf->st_mode)) {
			printf("/");
		} else if (show_slash > 1 && S_ISLNK(statbuf->st_mode)) {
			if (!hide_link) printf("@");
		} else if (show_slash > 1 && S_ISFIFO(statbuf->st_mode)) {
			printf("|");
		} else if (show_slash > 1 && (statbuf->st_mode & 0111)) {
			printf("*");
		} else if (show_slash > 1 && S_ISSOCK(statbuf->st_mode)) {
			printf("=");
		} else {
			printf(" ");
		}
	}
}

static void print_entry(struct tfile * file, int *colwidth) {
	prefixes(colwidth,file);

	const char * ansi_color_str = ls_color_str(file->name, &file->statbuf);

	if (S_ISLNK(file->statbuf.st_mode)) {
		if (file->lstatres && LS_C(ORPHAN)) {
			ansi_color_str = LS_C(ORPHAN);
		} else if (!file->lstatres && use_sym_target) {
			ansi_color_str = ls_color_str(file->link, &file->statbufl);
		}
	}

	/* Print the file name */
	if (use_color && ansi_color_str) {
		printf("%s%s%s%s%s", LS_C(LEFT), ansi_color_str, LS_C(RIGHT), file->name, LS_C(END));
	} else {
		printf("%s", file->name);
	}

	/* These classifiers are accounted for in width calculations separately */
	print_classifier(&file->statbuf, 0);

	/* Pad the rest of the column */
	for (int rem = colwidth[0] - display_width_of_string(file->name); rem > 0; rem--) {
		printf(" ");
	}
}

static int print_username(char * _out, size_t avail, int uid) {

	struct passwd * p = getpwuid(uid);
	int out = 0;

	if (p) {
		out = snprintf(_out, avail, "%s", p->pw_name);
	} else {
		out = snprintf(_out, avail, "%d", uid);
	}

	endpwent();

	return out;
}

static int print_human_readable_size(char * _out, size_t avail, size_t s) {
	if (s >= 1<<20) {
		size_t t = s / (1 << 20);
		return snprintf(_out, avail, "%d.%1dM", (int)t, (int)(s - t * (1 << 20)) / ((1 << 20) / 10));
	} else if (s >= 1<<10) {
		size_t t = s / (1 << 10);
		return snprintf(_out, avail, "%d.%1dK", (int)t, (int)(s - t * (1 << 10)) / ((1 << 10) / 10));
	} else {
		return snprintf(_out, avail, "%d", (int)s);
	}
}

static void update_column_widths(int * widths, struct tfile * file) {
	int n;

	/* Links */
	n = snprintf(NULL, 0, "%d", file->statbuf.st_nlink);
	if (n > widths[0]) widths[0] = n;

	/* User */
	n = print_username(NULL, 0, file->statbuf.st_uid);
	if (n > widths[1]) widths[1] = n;

	/* Group */
	n = print_username(NULL, 0, file->statbuf.st_gid);
	if (n > widths[2]) widths[2] = n;

	/* File size */
	if (human_readable) {
		n = print_human_readable_size(NULL, 0, file->statbuf.st_size);
	} else {
		n = snprintf(NULL, 0, "%d", (int)file->statbuf.st_size);
	}
	if (n > widths[3]) widths[3] = n;
}

static void print_entry_long(int * widths, int * colwidth, struct tfile * file) {
	prefixes(colwidth, file);

	const char * ansi_color_str = ls_color_str(file->name, &file->statbuf);
	mode_t mode = file->statbuf.st_mode;

	/* file permissions */
	if (S_ISLNK(mode))       { printf("l"); }
	else if (S_ISCHR(mode))  { printf("c"); }
	else if (S_ISBLK(mode))  { printf("b"); }
	else if (S_ISDIR(mode))  { printf("d"); }
	else { printf("-"); }
	printf("%c", (mode & S_IRUSR) ? 'r' : '-');
	printf("%c", (mode & S_IWUSR) ? 'w' : '-');
	printf("%c", (mode & S_IXUSR) ? ((mode & S_ISUID) ? 's' : 'x') : ((mode & S_ISUID) ? 'S' : '-'));
	printf("%c", (mode & S_IRGRP) ? 'r' : '-');
	printf("%c", (mode & S_IWGRP) ? 'w' : '-');
	printf("%c", (mode & S_IXGRP) ? ((mode & S_ISGID) ? 's' : 'x') : ((mode & S_ISGID) ? 'S' : '-'));
	printf("%c", (mode & S_IROTH) ? 'r' : '-');
	printf("%c", (mode & S_IWOTH) ? 'w' : '-');
	printf("%c", (mode & S_IXOTH) ? ((mode & S_ISVTX) ? 't' : 'x') : ((mode & S_ISVTX) ? 'T' : '-'));

	printf( " %*d ", widths[0], file->statbuf.st_nlink); /* number of links, not supported */

	char tmp[100];
	print_username(tmp, 100, file->statbuf.st_uid);
	printf("%-*s ", widths[1], tmp);
	print_username(tmp, 100, file->statbuf.st_gid);
	printf("%-*s ", widths[2], tmp);

	if (human_readable) {
		print_human_readable_size(tmp, 100, file->statbuf.st_size);
		printf("%*s ", widths[3], tmp);
	} else {
		printf("%*d ", widths[3], (int)file->statbuf.st_size);
	}

	char time_buf[80];
	struct tm * timeinfo = localtime((time_t*)&file->statbuf.st_mtime);
	if (timeinfo->tm_year == this_year) {
		strftime(time_buf, 80, "%b %d %H:%M", timeinfo);
	} else {
		strftime(time_buf, 80, "%b %d  %Y", timeinfo);
	}
	printf("%s ", time_buf);

	if (S_ISLNK(file->statbuf.st_mode)) {
		if (file->lstatres && LS_C(ORPHAN)) {
			ansi_color_str = LS_C(ORPHAN);
		} else if (!file->lstatres && use_sym_target) {
			ansi_color_str = ls_color_str(file->link, &file->statbufl);
		}
	}

	/* Print the file name */
	if (use_color && ansi_color_str) {
		printf("%s%s%s%s%s", LS_C(LEFT), ansi_color_str, LS_C(RIGHT), file->name, LS_C(END));
	} else {
		printf("%s", file->name);
	}
	print_classifier(&file->statbuf, 1);
	if (S_ISLNK(file->statbuf.st_mode) && *file->link) {
		printf(" -> ");
		const char * s = file->lstatres == 0 ? ls_color_str(file->link, &file->statbufl) : LS_C(MISS);
		if (use_color && s) {
			printf("%s%s%s%s%s", LS_C(LEFT), s, LS_C(RIGHT), file->link, LS_C(END));
		} else {
			printf("%s", file->link);
		}
		print_classifier(&file->statbufl, 1);
	}

	printf("\n");
}

#define X_S "\033[3m"
#define X_E "\033[0m"
static int show_help(int argc, char * argv[]) {
	fprintf(stderr,
			"%s - list files\n"
			"\n"
			"usage: %s [-aAfFhiklpsxC1] [" X_S "path" X_E "...]\n"
			"\n"
			" -a  --all             " X_S "list all files (including . files)" X_E "\n"
			" -A  --almost-all      " X_S "like -a, but still hide . and .." X_E "\n"
			" -f                    " X_S "display entries in original order" X_E "\n"
			" -F  --classify        " X_S "show characters after files depending on type" X_E "\n"
			" -h  --human-readable  " X_S "human-readable file sizes" X_E "\n"
			" -H                    " X_S "follow symlinks to directories on the command line" X_E "\n"
			"                       " X_S "(this is the default unless -l or -F is given)" X_E "\n"
			" -i  --inode           " X_S "display inode number before name" X_E "\n"
			" -k  --kibibytes       " X_S "(no-op; block size is always 1024)" X_E "\n"
			" -l                    " X_S "use a long listing format" X_E "\n"
			" -L  --dereference     " X_S "follow symbolic links" X_E "\n"
			" -p                    " X_S "show a / after directory names" X_E "\n"
			" -P                    " X_S "never follow links; cancel -L and -H" X_E "\n"
			" -s  --size            " X_S "show size (in block) before file names" X_E "\n"
			" -x                    " X_S "sort entries across columns instead of down" X_E "\n"
			" -C                    " X_S "format output by columns (default)" X_E "\n"
			" --color[=when]        " X_S "specify when to enable color output" X_E "\n"
			" --help                " X_S "show this help text" X_E "\n"
			"\n", argv[0], argv[0]);
	return 0;
}

static int show_usage(int argc, char * argv[]) {
	fprintf(stderr,
			"usage: %s [-aAfFhiklpsxC1] [" X_S "path" X_E "...]\n"
			" or try '%s --help' for more information\n",
			argv[0], argv[0]);
	return 1;
}

static void display_tfiles(struct tfile ** ents_array, int numents, int print_total) {
	int ent_max_len[4] = {0,0,0,0};
	size_t total_blocks = 0;
	for (int i = 0; i < numents; i++) {
		ent_max_len[0] = MAX(ent_max_len[0], display_width_of_string(ents_array[i]->name));
		if (show_inode) ent_max_len[1] = MAX(ent_max_len[1], snprintf(NULL,0,"%zu ",ents_array[i]->statbuf.st_ino));
		if (show_size) {
			size_t size = as_1k_blocks(ents_array[i]->statbuf.st_size);
			if (human_readable) {
				ent_max_len[2] = MAX(ent_max_len[2], print_human_readable_size(NULL, 0, size * 1024) + 1);
			} else {
				ent_max_len[2] = MAX(ent_max_len[2], snprintf(NULL,0,"%zu ",size));
			}
		}
		total_blocks += as_1k_blocks(ents_array[i]->statbuf.st_size);
	}
	int total_width = 0;
	for (size_t i = 0; i < sizeof(ent_max_len) / sizeof(*ent_max_len); ++i) {
		total_width += ent_max_len[i];
	}
	if (show_slash) total_width += 1;

	if (print_total) {
		printf("total %zu\n", total_blocks);
	}

	if (long_mode) {
		int widths[4] = {0,0,0,0};
		for (int i = 0; i < numents; i++) {
			update_column_widths(widths, ents_array[i]);
		}
		for (int i = 0; i < numents; i++) {
			print_entry_long(widths, ent_max_len, ents_array[i]);
		}
	} else {
		/* Determine the gridding dimensions */
		int col_ext = total_width + min_col_spacing;
		int cols = ((term_width + min_col_spacing) / col_ext);
		if (cols == 0) cols = 1;

		/* Print the entries */
		if (columns) {
			int rows = numents / cols;
			if (rows * cols < numents) rows++;
			for (int i = 0; i < rows; ++i) {
				print_entry(ents_array[i], ent_max_len);
				for (int j = 1; j < cols; ++j) {
					if (i + j * rows >= numents) break;
					for (int t = 0; t < min_col_spacing; ++t) printf(" ");
					print_entry(ents_array[i + j * rows], ent_max_len);
				}
				printf("\n");
			}
		} else {
			for (int i = 0; i < numents;) {
				print_entry(ents_array[i++], ent_max_len);
				for (int j = 0; (i < numents) && (j < (cols-1)); j++) {
					for (int t = 0; t < min_col_spacing; ++t) printf(" ");
					print_entry(ents_array[i++], ent_max_len);
				}
				printf("\n");
			}
		}
	}
}

static int display_dir(char * p) {
	/* Open the directory */
	DIR * dirp = opendir(p);
	if (dirp == NULL) {
		return 2;
	}

	if (print_dir) {
		printf("%s:\n", p);
	}

	/* Read the entries in the directory */
	list_t * ents_list = list_create();

	struct dirent * ent = readdir(dirp);
	for (; ent; ent = readdir(dirp)) {
		if (ent->d_name[0] == '.' && !show_hidden) continue;
		if (show_hidden != 1 && !strcmp(ent->d_name, ".")) continue;
		if (show_hidden != 1 && !strcmp(ent->d_name, "..")) continue;

		struct tfile * f = malloc(sizeof(struct tfile));

		f->name = strdup(ent->d_name);

		char tmp[strlen(p)+strlen(ent->d_name)+2];
		sprintf(tmp, "%s/%s", p, ent->d_name);
		(follow_all_links ? stat : lstat)(tmp, &f->statbuf);
		if (S_ISLNK(f->statbuf.st_mode)) {
			f->lstatres = stat(tmp, &f->statbufl);
			f->link = malloc(f->statbuf.st_size + 1);
			ssize_t len = readlink(tmp, f->link, f->statbuf.st_size);
			if (len >= 0) f->link[len] = '\0';
			else {
				warn("cannot read symbolic link '%s'", tmp);
				f->link[0] = '\0';
			}
		}

		list_insert(ents_list, (void *)f);
	}
	closedir(dirp);

	/* Now, copy those entries into an array (for sorting) */
	if (!ents_list->length) return 0;

	struct tfile ** file_arr = malloc(sizeof(struct tfile *) * ents_list->length);
	int index = 0;
	foreach(node, ents_list) {
		file_arr[index++] = (struct tfile *)node->value;
	}

	list_free(ents_list);

	if (!in_order) qsort(file_arr, index, sizeof(struct tfile *), filecmp_notypesort);

	display_tfiles(file_arr, index, long_mode || show_size); /* || a few other things as well we don't have yet */

	free(file_arr);

	return 0;
}

static void setup_colors(void) {
	ls_colors_init();

	if (LS_C(SYM) && !strcmp(LS_C(SYM), "target")) {
		use_sym_target = 1;
		LS_C(SYM) = "1;36";
	}

	char * ls_colsep = getenv("LS_COLSEP");
	if (ls_colsep) {
		min_col_spacing = atoi(ls_colsep);
	}
}

int main (int argc, char * argv[]) {
	int assume_no_follow = 0;
	char * p = ".";

	static struct option long_opts[] = {
		{"help", no_argument, 0, '-'},
		{"all", no_argument, 0, 'a'},
		{"almost-all", no_argument, 0, 'A'},
		{"classify", no_argument, 0, 'F'},
		{"dereference", no_argument, 0, 'L'},
		{"human-readable", no_argument, 0, 'h'},
		{"inode", no_argument, 0, 'i'},
		{"kibibytes", no_argument, 0, 'k'},
		{"size", no_argument, 0, 's'},
		{"color", optional_argument, 0, 1000},
		{0,0,0,0},
	};

	stdout_is_tty = isatty(STDOUT_FILENO);
	use_color = stdout_is_tty; /* we default to 'auto' */

	int opt, index;
	while ((opt = getopt_long(argc, argv, "aAfFhHiklLpPsxC1?", long_opts, &index)) != -1) {
		switch (opt) {
			case 'a':
				show_hidden = 1;
				break;
			case 'A':
				show_hidden = 2;
				break;
			case 'f':
				in_order = 1;
				break;
			case 'F':
				show_slash = 2;
				assume_no_follow = 1;
				break;
			case 'h':
				human_readable = 1;
				break;
			case 'i':
				show_inode = 1;
				break;
			case 'k':
				/* no op, we always use 1024 like GNU */
				break;
			case 'l':
				long_mode = 1;
				assume_no_follow = 1;
				break;
			case 'p':
				show_slash = 1;
				break;
			case 's':
				show_size = 1;
				break;
			case 'H':
				follow_cmdline_dir_links = 1;
				break;
			case 'L':
				follow_cmdline_dir_links = 1;
				follow_all_links = 1;
				assume_no_follow = 0;
				break;
			case 'P':
				follow_cmdline_dir_links = 0;
				follow_all_links = 0;
				assume_no_follow = 1;
				break;

			/* TODO These two should also force the multi-column
			 * display even when stdout is not a TTY... */
			case 'x':
				long_mode = 0;
				columns = 0;
				term_width = 79;
				break;
			case 'C':
				long_mode = 0;
				columns = 1;
				term_width = 79;
				break;

			case '1':
				one_column = 1;
				break;

			case 1000:
				if (optarg) {
					if (!strcmp(optarg, "never") || !strcmp(optarg,"n")) {
						use_color = 0;
					} else if (!strcmp(optarg, "auto")) {
						use_color = stdout_is_tty;
					} else if (!strcmp(optarg, "always")) {
						use_color = 1;
					} else {
						errx(1, "--color= must be one of 'never', 'auto', or 'always'");
					}
				} else {
					use_color = 1; /* --color is equivalent to --color=always */
				}
				break;

			case '-':
				if (index == 0) return show_help(argc, argv);
				/* fallthrough */
			case '?':
				return show_usage(argc, argv);
		}
	}

	if (optind < argc) p = argv[optind];
	if (optind + 1 < argc) print_dir = 1;

	if (!assume_no_follow) follow_cmdline_dir_links = 1;

	if (long_mode) {
		struct tm * timeinfo;
		struct timeval now;
		gettimeofday(&now, NULL); //time(NULL);
		timeinfo = localtime((time_t *)&now.tv_sec);
		this_year = timeinfo->tm_year;
	}

	if (stdout_is_tty && !one_column) {
		struct winsize w;
		ioctl(1, TIOCGWINSZ, &w);
		term_width = w.ws_col;
		term_width -= 1; /* And this just helps clean up our math */
	}

	setup_colors();

	int out = 0;

	if (argc == 1 || optind == argc) {
		if (display_dir(p) == 2) {
			warn("cannot access '%s'", p);
		}
	} else {
		list_t * files = list_create();
		while (p) {
			struct tfile * f = malloc(sizeof(struct tfile));

			f->name = p;
			int t = (follow_all_links ? stat : lstat)(p, &f->statbuf);

			if (t < 0) {
				warn("cannot access '%s'", p);
				free(f);
				out = 2;
			} else {
				if (S_ISLNK(f->statbuf.st_mode)) {
					f->lstatres = stat(p, &f->statbufl);
					if (follow_cmdline_dir_links && S_ISDIR(f->statbufl.st_mode)) {
						memcpy(&f->statbuf, &f->statbufl, sizeof(struct stat));
						goto _insert;
					}
					f->link = malloc(f->statbuf.st_size + 1);
					ssize_t len = readlink(p, f->link, f->statbuf.st_size);
					if (len >= 0) f->link[len] = '\0';
					else {
						warn("cannot read symbolic link '%s'", p);
						f->link[0] = '\0';
					}
				}
_insert:
				list_insert(files, f);
			}

			optind++;
			if (optind >= argc) p = NULL;
			else p = argv[optind];
		}

		if (!files->length) {
			/* No valid entries */
			return out;
		}

		struct tfile ** file_arr = malloc(sizeof(struct tfile *) * files->length);
		int index = 0;
		foreach(node, files) {
			file_arr[index++] = (struct tfile *)node->value;
		}

		list_free(files);

		qsort(file_arr, index, sizeof(struct tfile *), filecmp);

		int first_directory = index;

		for (int i = 0; i < index; ++i) {
			if (S_ISDIR(file_arr[i]->statbuf.st_mode)) {
				first_directory = i;
				break;
			}
		}

		if (first_directory) {
			display_tfiles(file_arr, first_directory, 0);
		}

		for (int i = first_directory; i < index; ++i) {
			if (i != 0) {
				printf("\n");
			}
			if (display_dir(file_arr[i]->name) == 2) {
				warn("cannot access '%s'", file_arr[i]->name);
			}
		}
	}

	return out;
}

